From 6cb18b38bbf60bc89783951f80ebfad29de14c02 Mon Sep 17 00:00:00 2001
From: Jan Kiszka <jan.kiszka@siemens.com>
Date: Sun, 1 Jan 2017 15:43:37 +0100
Subject: [PATCH 066/184] ivshmem-net: Switch to reset state on each net stop
 and on driver removal

Improves the state signaling to the remote side after ifconfig down and
driver removal.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>
---
 drivers/net/ivshmem-net.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ivshmem-net.c b/drivers/net/ivshmem-net.c
index 593a84e12..afff08044 100644
--- a/drivers/net/ivshmem-net.c
+++ b/drivers/net/ivshmem-net.c
@@ -555,6 +555,8 @@ static void ivshm_net_do_stop(struct net_device *ndev)
 {
 	struct ivshm_net *in = netdev_priv(ndev);
 
+	ivshm_net_set_state(in, IVSHM_NET_STATE_RESET);
+
 	if (!test_and_clear_bit(IVSHM_NET_FLAG_RUN, &in->flags))
 		return;
 
@@ -594,7 +596,6 @@ static void ivshm_net_state_change(struct work_struct *work)
 		} else {
 			netif_carrier_off(ndev);
 			ivshm_net_do_stop(ndev);
-			ivshm_net_set_state(in, IVSHM_NET_STATE_RESET);
 		}
 		break;
 	}
@@ -900,6 +901,8 @@ static void ivshm_net_remove(struct pci_dev *pdev)
 	struct net_device *ndev = pci_get_drvdata(pdev);
 	struct ivshm_net *in = netdev_priv(ndev);
 
+	writel(IVSHM_NET_STATE_RESET, &in->ivshm_regs->lstate);
+
 	if (in->using_msix)  {
 		free_irq(in->msix.vector, ndev);
 		pci_disable_msix(pdev);
-- 
2.39.2

